lib/membrane/video_compositor/video_compositor.ex

```diff
@@ -0,0 +1,431 @@
+    state = remove_output(state, output_ref)
+    {[remove_child: [{:rtp_receiver, pad_id}, {:upd_source, pad_id}]], state}
+  end
+
+  @impl true
+  def handle_parent_notification(:start_composing, _ctx, state = %State{}) do
+    :ok = VcReq.start_composing(state.vc_port)
+    {[], state}
+  end
+
+  @impl true
+  def handle_parent_notification({:vc_request, request_body}, _ctx, state = %State{}) do
+    case VcReq.send_custom_request(request_body, state.vc_port) do
+      {:ok, response} ->
+        if response.status != 200 do
+          Membrane.Logger.error(
+            "Request\n#{inspect(request_body)}\nfailed with error:\n#{inspect(response.body)}"
+          )
+        end
+
+        {[notify_parent: {:vc_request_response, request_body, response, State.ctx(state)}], state}
+
+      {:error, err} ->
+        Membrane.Logger.error("Request: #{request_body} failed. Error: #{err}.")
+        {[], state}
+    end
+  end
+
+  @impl true
+  def handle_parent_notification(_notification, _ctx, state = %State{}) do
+    {[], state}
+  end
+
+  @impl true
+  def handle_child_notification(
+        {:new_rtp_stream, ssrc, _pt, _ext},
+        {:rtp_receiver, pad_id},
+        _ctx,
+        state = %State{}
+      ) do
+    %OutputState{resolution: %Resolution{width: output_width, height: output_height}} =
+      state.outputs
+      |> Enum.find(fn %OutputState{pad_ref: Pad.ref(:output, id)} -> pad_id == id end)
+
+    output_stream_format = %Membrane.H264{
+      framerate: {state.framerate, 1},
+      alignment: :nalu,
+      stream_structure: :annexb,
+      width: output_width,
+      height: output_height
+    }
+
+    spec =
+      get_child({:rtp_receiver, pad_id})
+      |> via_out(Pad.ref(:output, ssrc), options: [depayloader: RTP.H264.Depayloader])
+      |> child({:output_processor, pad_id}, %Membrane.VideoCompositor.OutputProcessor{
+        output_stream_format: output_stream_format
+      })
+      |> bin_output(Pad.ref(:output, pad_id))
+
+    {[spec: spec], state}
+  end
+
+  @impl true
+  def handle_child_notification(_msg, _child, _ctx, state) do
+    {[], state}
+  end
+
+  @spec start_vc_server(port_number()) :: :ok
+  defp start_vc_server(vc_port) do
+    architecture = system_architecture() |> Atom.to_string()
+
+    vc_app_path =
+      File.cwd!()
+      |> Path.join("video_compositor_app/#{architecture}/video_compositor/video_compositor")
+
+    spawn(fn ->
+      Rambo.run(vc_app_path, [], env: %{"MEMBRANE_VIDEO_COMPOSITOR_API_PORT" => "#{vc_port}"})
+    end)
+
+    started? =
+      0..50
+      |> Enum.reduce_while(false, fn _i, _acc ->
+        sleep_time_ms = 100
+        :timer.sleep(sleep_time_ms)
+
+        case VcReq.send_custom_request(%{}, vc_port) do
+          {:ok, _} -> {:halt, true}
+          {:error, _} -> {:cont, false}
+        end
+      end)
+
+    unless started? do
+      raise "Failed to startup and connect to VideoCompositor server."
+    end
+
+    :ok
+  end
```

Comment:
`VcReq.send_custom_request` is the only function from `VcReq` module, that returns an `:ok` tuple if `response.status != 200`. You check the `response.status` in line 251, but not in `case` in line 325.  Are you sure, that this is ok? For me it looks little inconsistent and opens a place for easy mistakes, so it would be good to refactor it

Reply:
Yep, this is basically pinging the VideoCompositor server to check whether it's already up. I'll make this more readable.